src/datepicker/calendar-header.js

```diff
@@ -66,45 +53,68 @@ function idToYearMonth(id) {
   return id.split('-').map(Number);
 }
 
-export default class CalendarHeader extends React.Component<
-  HeaderPropsT,
+export default class CalendarHeader<T = Date> extends React.Component<
+  HeaderPropsT<T>,
   {isMonthYearDropdownOpen: boolean, isFocusVisible: boolean},
 > {
   static defaultProps = {
-    date: new Date(),
+    adapter: dateFnsAdapter,
     locale: null,
     maxDate: null,
     minDate: null,
     onYearChange: () => {},
     overrides: {},
   };
 
-  state = {isMonthYearDropdownOpen: false, isFocusVisible: false};
+  dateHelpers: DateHelpers<T>;
+
+  constructor(props: HeaderPropsT<T>) {
+    super(props);
+    this.dateHelpers = new DateHelpers(props.adapter);
+  }
+
+  state = {
+    isMonthYearDropdownOpen: false,
+    isFocusVisible: false,
+  };
+
+  getDateProp: () => T = () => {
```

Comment:
The old pattern of creating a javascript date as a default prop isn't going to work with the generic date type. 

The simplest solution would be to assign `dateHelpers.date()` as the default instead, but because `dateHelpers` isn't available until after the constructor is called, moving the default handling into a function seemed like the best bet.